### PR TITLE
CORE-648 fixes based on experience implementing an HSM

### PR DESCRIPTION
**Pull request by alexey-kadyrov-r3**
* CryptoService should report key schemes and supported signature specs for each key scheme as some of the EC curves are linked to which digest can be used by signing (eg. spec256r1 should use SHA256withECDSA)
* Supporting the throttling on the platform level
* Instead of using optional interfaces for additional functionality (like delete) it should be single interface coupled with reporting of capabilities (due decorators to handle common fault conditions and throttling)
* Signature spec made polymorphic to simply use in the application layer

### CODE REVIEW

File: cipher-suite/src/main/kotlin/net/corda/v5/cipher/suite/CryptoService.kt [hunk 1]
```diff
@@ -10,41 +10,38 @@ import net.corda.v5.crypto.exceptions.CryptoServiceException
  * Note about key aliases. Corda always uses single alias to identify a key pair however some HSMs need separate
  * aliases for public and private keys, in such cases their names have to be derived from the single key pair alias.
  * It could be suffixes or whatever internal naming scheme is used.
- * Also note that's not required to keep a public
- * key in the HSM as that will be kept by the upstream Crypto Services.
+ *
+ * Also note that's not required to keep a public key in the HSM as that will be kept by the upstream Crypto Services.
```
Comment (rick-r3) [inline cipher-suite/src/main/kotlin/net/corda/v5/cipher/suite/CryptoService.kt]: that's not -> that it is not

File: cipher-suite/src/main/kotlin/net/corda/v5/cipher/suite/CryptoService.kt [hunk 2]
```diff
@@ -10,41 +10,38 @@ import net.corda.v5.crypto.exceptions.CryptoServiceException
...
+ *
+ * Exception handling.
+ *
+ * As the service instances are decorated with decorators taking care about the timeout, throttling handling,
```
Comment (rick-r3) [inline cipher-suite/src/main/kotlin/net/corda/v5/cipher/suite/CryptoService.kt]: care about -> care of

File: cipher-suite/src/main/kotlin/net/corda/v5/cipher/suite/CryptoService.kt [hunk 3]
```diff
@@ -10,41 +10,38 @@ import net.corda.v5.crypto.exceptions.CryptoServiceException
...
+ * exception normalisation the exceptions listed for each method are as expected be thrown by the decorators.
```
Comment (rick-r3) [inline cipher-suite/src/main/kotlin/net/corda/v5/cipher/suite/CryptoService.kt]: exception normalisation -> and exception normalisation
expected be thrown -> expected to be thrown

File: cipher-suite/src/main/kotlin/net/corda/v5/cipher/suite/CryptoService.kt [hunk 4]
```diff
@@ -10,41 +10,38 @@ import net.corda.v5.crypto.exceptions.CryptoServiceException
...
+ *
+ * The service should throw the most appropriate exception as it sees it, e.g. if the input parameters are wrong it
+ * should be [IllegalArgumentException] or if the internal state is wrong for an operation then the most appropriate
+ * exception would be [IllegalStateException].
+ *
+ * If service encountered throttling situation then it should throw one
+ * of concrete implementations of [CSLThrottlingException] (such as [CSLExponentialThrottlingException]) so the
+ * library can handle it appropriately.
+ *
+ * The next exceptions are retried - [TimeoutException], [CryptoServiceLibraryException] (with the isRecoverable flage
```
Comment (rick-r3) [inline cipher-suite/src/main/kotlin/net/corda/v5/cipher/suite/CryptoService.kt]: The next exceptions -> The following exceptions

**Review by rick-r3 — DISMISSED** [review]

**Review by mnesbit — APPROVED** [review]
ok

**Pull request merged**